src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxResolvingClassLoader.java

```diff
@@ -39,14 +40,19 @@
             return this.getClass().getClassLoader().loadClass(name);
         } else {
             ClassLoader parentLoader = getParent();
-            Class<?> c = load(parentClassCache, name, parentLoader, () -> {
+            Optional<WeakReference<Class<?>>> ref = load(parentClassCache, name, parentLoader, () -> {
                 try {
-                    return Optional.of(parentLoader.loadClass(name));
+                    return Optional.of(new WeakReference<>(parentLoader.loadClass(name)));
                 } catch (ClassNotFoundException x) {
                     return Optional.absent();
                 }
-            }).orNull();
-            if (c != null) {
+            });
+            if (ref.isPresent()) {
+                Class<?> c = ref.get().get();
+                if (c == null) {
+                    // Probably impossible for the class to be collected if its loader was not, but just in case…
```

Comment:
Maybe worth a FINE-level log message? Would be nice to have an easy way to specify behavior like "log at FINE initially, but if you see this more than X times over an X-TimeUnit period, log at WARNING" in case something goes wrong and this is called way more often than expected.

Reply:
I guess it could be an assertion, but it did not seem worth the bother.